### PR TITLE
ENH: analytic version of arma_acovf

### PR DESCRIPTION
**Pull request by ChadFulton**
Updates `arima_process.arma_acovf` to compute the autocovariances using the analytical approach of Brockwell and Davis (1991).

- Somewhat faster in ~~most~~ some cases, much faster in cases with high persistence
- Improves precision
- Adds additional unit tests

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Do you have an example where the new implementation outperforms (in speed or accuracy or ...) the old one?

**Comment by josef-pkt** [comment]
My old version makes a detour via a infinite series that needs truncation, so there is a truncation error if the truncation happens too soon, or wasteful computation if truncation leaves a large number of terms.
The old version might have a speed advantage for large m (convolution versus lstsq), but in all other cases the new version should be faster, besides being more accurate.

**Comment by josef-pkt** [comment]
I'm not sure anymore whether there is a truncation error for acovf at small lags

```
ir = arma_impulse_response(ar, ma, leads=nobs_ir)
acovf = np.correlate(ir, ir, 'full')[len(ir) - 1:]
```

This looks similar to the forecast variance on the mailing list (except for the missing cumsum?), and it should be possible to compute the forecast covariance at different horizons/steps to get the joint distribution and a simultaneous confidence interval.

Computationally the new version is better and LGTM, 
For the theoretical interpretation, I liked the old version.

File: statsmodels/tsa/tests/test_arima_process.py [hunk 1]
```diff
@@ -73,7 +73,25 @@ def test_arma_acf():
     # rep 2: manually
     acovf = np.array([1. * sigma * phi ** i / (1 - phi ** 2) for i in range(N)])
     rep2 = acovf / (1. / (1 - phi ** 2))
-    assert_almost_equal(rep1, rep2, 8)  # 8 is max precision here
+    assert_allclose(rep1, rep2)
+
+    # Test other cases, against R's ARMAacf
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/test_arima_process.py]:
```suggestion
def test_arma_acf_compare_R_ARMAacf():
    # Test other cases, against R's ARMAacf; GH#5324
```

File: statsmodels/tsa/tests/test_arima_process.py [hunk 2]
```diff
@@ -155,8 +173,8 @@ def test_armafft():
             arma = ArmaFft(ar, ma, 20)
             ac1 = arma.invpowerspd(1024)[:10]
             ac2 = arma.acovf(10)[:10]
-            assert_almost_equal(ac1, ac2, decimal=7,
-                                err_msg='acovf not equal for %s, %s' % (ar, ma))
+            assert_allclose(ac1, ac2, atol=1e-15,
+                            err_msg='acovf not equal for %s, %s' % (ar, ma))
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/test_arima_process.py]: Are the actual values of ac1 and ac2 big enough that atol=1e-15 is decently tight?

File: statsmodels/tsa/tests/test_arima_process.py [hunk 3]
```diff
@@ -155,8 +173,8 @@ def test_armafft():
             arma = ArmaFft(ar, ma, 20)
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/test_arima_process.py]: This test should be parametrized

File: statsmodels/tsa/tests/test_arima_process.py [hunk 4]
```diff
@@ -59,8 +60,7 @@ def test_arma_acovf_persistent():
     corrs = .9995**np.arange(10)
     expected = sig2*corrs
     assert_equal(res.ndim, 1)
-    assert_allclose(res, expected, atol=1e-6)
-    # atol=7 breaks at .999, worked at .995
+    assert_allclose(res, expected)
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/test_arima_process.py]: Does the change in this PR improve the match?  (i.e. the resolve the comment above)

File: statsmodels/tsa/arima_process.py [hunk 5]
```diff
@@ -93,6 +93,8 @@ def arma_acovf(ar, ma, nobs=10):
         coefficient for moving-average lag polynomial, including zero lag
     nobs : int
         number of terms (lags plus zero lag) to include in returned acovf
+    sigma2 : float
+        Variance of the innovation term.
```
Comment (jbrockmendel) [inline statsmodels/tsa/arima_process.py]: require strictly positive?

File: statsmodels/tsa/tests/results/arma_acov.R [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+# Custom example 6
+ar_params = c(1, -0.25, 0.1, -0.05)
+ma_params = c(1, 0.2)
+print(ARMAacf(ar_params, ma_params, lag.max=9))
+
+# Custom example 7
+ar_params = c(1, -0.25, 0.1, -0.05, 0.02)
+ma_params = c(1, 0.2)
+print(ARMAacf(ar_params, ma_params, lag.max=9))
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/results/arma_acov.R]: Is there a manual step to create results/arma_acov_results.py from the output of this script?  Can that step be automated?

File: statsmodels/tsa/tests/results/arma_acov_results.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
+bd_example_3_3_2 = [
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/results/arma_acov_results.py]: Comment describing where these numbers come from, how to reproduce

**Comment by ChadFulton** [comment]
I also forgot to mention that an alternative to this PR is to keep both approaches around in some form, so if that's preferable let me know.

**Comment by josef-pkt** [comment]
I thought about adding a `method` argument, but I don't think it is worth it. 
AFAIK, the new code is computationally much better than the old, and doesn't need the guessing for the truncation.

(If there is interest, then parking the simplified old version in an example or test case could be useful just for documenting the theoretical relationship. Sometimes I like to keep an in most cases computationally inferior version around just as documentation or for quick experiments, but this is not really my area anymore.)

**Comment by ChadFulton** [comment]
Thanks for the review @jbrockmendel and the comments @josef-pkt.

File: statsmodels/tsa/tests/test_arima_process.py (hunk 1 above)
Comment (ChadFulton) [reply statsmodels/tsa/tests/test_arima_process.py]: Why the reference to the github PR?

File: statsmodels/tsa/tests/test_arima_process.py (hunk 4 above)
Comment (ChadFulton) [reply statsmodels/tsa/tests/test_arima_process.py]: Yes, all of the precision improvements are due to the analytic compuation.

File: statsmodels/tsa/tests/test_arima_process.py (hunk 3 above)
Comment (ChadFulton) [reply statsmodels/tsa/tests/test_arima_process.py]: Noted :)

**Comment by ChadFulton** [comment]
> (If there is interest, then parking the simplified old version in an example or test case could be useful just for documenting the theoretical relationship. Sometimes I like to keep an in most cases computationally inferior version around just as documentation or for quick experiments, but this is not really my area anymore.)

This is a good idea, I will make a test case that uses it. Thanks!

File: statsmodels/tsa/tests/test_arima_process.py (hunk 1 above)
Comment (jbrockmendel) [reply statsmodels/tsa/tests/test_arima_process.py]: Habit.  Avoids the need to look up git history.

File: statsmodels/tsa/tests/results/arma_acov.R (hunk 6 above)
Comment (jbrockmendel) [reply statsmodels/tsa/tests/results/arma_acov.R]: Don’t bother; only if there were an obvious way to do it.

**Comment by ChadFulton** [comment]
Green as it gets right now (it looks like some deprecation warnings about `np.matrix` from Numpy are flooding the Travis log, and causing it to error out before finishing on three of the runs).

**Comment by jbrockmendel** [comment]
If you're interested in clearing warnings out from the logs: #5260, #5204 are ready

File: statsmodels/tsa/arima_process.py [hunk 8]
```diff
@@ -104,35 +106,42 @@ def arma_acovf(ar, ma, nobs=10):
     arma_acf
     acovf
 
+    References
+    ----------
+    Brockwell, Peter J., and Richard A. Davis. 2009.
```
Comment (bashtage) [inline statsmodels/tsa/arima_process.py]: Too late now, but ideally would be in rst reference style `.. [*] Brockwell`...

**Pull request merged**